File: problems/0003-longest-substring-without-repeating-characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        """
        Problem: 0003 - Longest Substring Without Repeating Characters.

        Given a string, find the length of the longest substring without repeating characters.

        Approach:
        - Use a sliding window (left and right pointers).
        - Expand the right pointer to include new characters.
        - If a duplicate is found, move the left pointer forward
          until the window contains only unique characters again.
        - Track the maximum window size during traversal.

        Time Complexity: O(n), where n is the length of the string.
        Space Complexity: O(min(n, m)), where m is the charset size (e.g., 128).
        """

        seen = set()        # Characters currently in the window.
        left = 0            # Left boundary of the window.
        max_len = 0         # Length of the longest valid window found.

        for right in range(len(s)):
            # Shrink the window from the left while s[right] is a duplicate.
            while s[right] in seen:
                seen.remove(s[left])    # Remove the leftmost character.
                left += 1               # Move the left boundary forward.

            seen.add(s[right])          # Add current character to the set.
            max_len = max(max_len, right - left + 1)  # Update max length.

        return max_len

File: problems/0003-longest-substring-without-repeating-characters/solution.cpp
#include <string>
#include <unordered_set>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        /*
        Problem: 0003 - Longest Substring Without Repeating Characters.

        Given a string, return the length of the longest substring without repeating characters.

        Approach:
        - Use a sliding window with two pointers: left and right.
        - Use an unordered_set to track characters currently in the window.
        - Expand the right pointer on each iteration.
        - If a duplicate character is found, shrink the window from the left
          until all characters are unique again.
        - Track and update the max window length.

        Time Complexity: O(n), where n is the length of the string.
        Space Complexity: O(min(n, m)), where m is the charset size.
        */

        unordered_set<char> seen;   // Set of unique characters in the window.
        int left = 0, max_len = 0;

        for (int right = 0; right < s.length(); ++right) {
            // While current character is a duplicate, shrink from the left.
            while (seen.find(s[right]) != seen.end()) {
                seen.erase(s[left]);   // Remove leftmost character.
                ++left;                // Move left pointer forward.
            }

            seen.insert(s[right]);     // Add new character to window.
            max_len = max(max_len, right - left + 1);  // Update max length.
        }

        return max_len;
    }
};

